### PR TITLE
Fix NSTimer Race Condition

### PR DESCRIPTION
**Pull request by sgoldberg-sfdc**
- Original code was developed to ensure thread is stopped on same thread it is created on
- This can also be accomplished by ensuring it is always run on the main thread (which in practice it most often was)
- If the timer had been created on a background thread this could result in a race condition between the performselector call and the thread terminating.

### CODE REVIEW

**Comment by sgoldberg-sfdc** [comment]
@bhariharan @khawkins

**Comment by bhariharan** [comment]
test me

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m [hunk 1]
```diff
@@ -932,39 +931,43 @@ - (void)resetWebUserAgent
 
 - (void)startRefreshFlowConnectionTimer
 {
-    self.refreshTimerThread = [NSThread currentThread];
+    if (![NSThread isMainThread]) {
+        __weak typeof(self) weakSelf = self;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m]: `dispatch_async()` will not capture `self`, and in this case it's possible to lose the reference to `self` as a result.  I'd remove the casting here.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m [hunk 2]
```diff
@@ -932,39 +931,43 @@ - (void)resetWebUserAgent
...
+                                                                          repeats:NO];
 }
 
 - (void)stopRefreshFlowConnectionTimer
 {
-    if (self.refreshFlowConnectionTimer != nil && self.refreshTimerThread != nil) {
-        [self performSelector:@selector(invalidateRefreshTimer) onThread:self.refreshTimerThread withObject:nil waitUntilDone:YES];
-        [self cleanupRefreshTimer];
+    if (![NSThread isMainThread]) {
+        __weak typeof(self) weakSelf = self;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m]: Same here.

**Comment by sgoldberg-sfdc** [comment]
Removed weak/strong dance per @khawkins

**Comment by bhariharan** [comment]
test me

**Comment by sgoldberg-sfdc** [comment]
All green!

**Comment by sgoldberg-sfdc** [comment]
@khawkins  and @bhariharan - On second thought, I think it would be better to replace this NStimer with a dispatch_after or dispatch_source_set_timer().

**Pull request merged**